docs/show-changed-rco.1
.\" show-changed-rco
.TH "show-installed" "1" "30 March 2011" "James Antill" ""
.SH "NAME"
show\-changed\-rco \- show changes in an RPM package
.SH "SYNOPSIS"
\fBshow\-changed\-rco\fP [options]
.SH "DESCRIPTION"
.PP
\fBshow\-changed\-rco\fP gives a compact description of the changes to a
packages Requires, Conflicts and Obsoletes data from the installed (or old) to
a specified rpm file.
.SH OPTIONS
.TP
.IP "\fB\-h, \-\-help\fP"
show this help message and exit
.IP "\fB\-C, \-\-cache\fP" 
Tells repoquery to run entirely from YUM cache - does not download any metadata
or update the cache. Queries in this mode can fail or give partial/incorrect
results if the cache isn't fully populated beforehand with eg "yum makecache".
.IP "\fB\-c <config file>, \-\-config=<config file>\fP"
Use alternative config file (default is /etc/yum.conf).

.IP "\fB\-\-repoid=<repo>\fP"
Specify which repository to query. Using this option disables all repositories
not explicitly enabled with \-\-repoid option (can be used multiple times). By
default repoquery uses whatever repositories are enabled in YUM configuration.
.IP "\fB\-\-enablerepo=<repo>\fP"
In addition to the default set, query the given additional repository, even if
it is disabled in YUM configuration.  Can be used multiple times.
.IP "\fB\-\-disablerepo=<repo>\fP"
Do not query the given repository, even if it is enabled in YUM
configuration.  Can be used multiple times.
.IP "\fB\-\-repofrompath=<repoid>,<path/url>\fP"
Specify a path or url to a repository (same path as in a baseurl) to add to
the repositories for this query. This option can be used multiple times. If
you want to view only the pkgs from this repository combine this with
\-\-repoid. The repoid for the repository is specified by <repoid>.

.IP "\fB\-\-old\-packages=<pkg>\fP"
Explicitly list the valid old packages to match the new packages against.
.IP "\fB\-\-ignore\-arch\fP"
Ignore arch. so you can compare foo-2.i686 to foo-1.x86_64.
.IP "\fB\-\-skip\-new\fP"
Only give output for packages which we've found an old package for.

.PP 
.SH "SEE ALSO"
.nf
.I yum.conf (5)
http://yum.baseurl.org/
.fi 

.PP 
.SH "AUTHORS"
.nf 
See the Authors file included with this program.
.fi 

.PP 
.SH "BUGS"
There are of course no bugs, but should you find any, you should first
consult the FAQ section on http://yum.baseurl.org/wiki/Faq and if unsuccessful
in finding a resolution contact the mailing list: yum-devel@lists.baseurl.org.
To file a bug use http://bugzilla.redhat.com for Fedora/RHEL/Centos
related bugs and http://yum.baseurl.org/report for all other bugs.

.fi


ChangeLog
2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* ChangeLog: updated ChangeLog

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-03  Shawn Starr <shawn.starr@rogers.com>

	* docs/yum-utils.1: Update yum-utils man page, to add new utils.

2011-07-25  James Antill <james@and.org>

	Merge branch 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils  * 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils: (3 commits)
	typo leaving .conf in ybc.conf.yumvar should just be ybc.yumvar
	...

2011-07-25  James Antill <james@and.org>

	* plugins/aliases/aliases: Add some new aliases, fix chkD.

2011-07-25  Seth Vidal <skvidal@fedoraproject.org>

	* yum-config-manager.py: typo leaving .conf in ybc.conf.yumvar
	should just be ybc.yumvar

2011-07-12  James Antill <james@and.org>

	* plugins/security/security.py: Also auto-apply --security when
	given --sec-severity for other commands.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add repoquery @group completion.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* repoquery.py: Fix "nitmes" typo in previous commit.

2011-07-06  James Antill <james@and.org>

	* repoquery.py: Fix typo

2011-06-29  James Antill <james@and.org>

	* repoquery.py: Allow @foo groups as a packages selector in
	repoquery.

2011-06-25  Ville Skyttä <ville.skytta@iki.fi>

	* plugins/aliases/aliases.py, plugins/security/security.py,
	plugins/versionlock/versionlock.py: Remove some unused local
	variables and imports flagged by pyflakes.

2011-06-24  Ville Skyttä <ville.skytta@iki.fi>

	* debuginfo-install.py, package-cleanup.py, yum-builddep.py,
	yum-complete-transaction.py: Check doUtilBuildTransaction return
	value, exit early on failure.  BZ 716267

2011-06-22  James Antill <james@and.org>

	* plugins/security/security.py: Try to be clever when sorting the
	security severities.

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: After (external) merge is
	complete, give user option to delete other file

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: Support meld as well as vimdiff
	for merging  When using meld, the left side has the "other file"
	i.e. the file with the .rpmnew or .rpmsave extension. The right side
	has the "final file" i.e. the target file into which changes have to
	be merged.  If the rpm action was to save the existing conf file,
	then the left side will be the original conf file (.rpmsave) and the
	right side will be the new rpm-based configuration. The user should
	merge any changes they want to keep from their old .rpmsave from
	left to right.  If the rpm action was to keep the existing conf
	file, then the left side will be the new conf file (.rpmnew) and
	right side will be the current local configuration. The user should
	merge any changes in the new rpm file into their local config from
	left to right.  Ideally these operations would be 3-way diffs
	between the local file and the new rpm config, using the original
	rpm config as the common merge base. However, the original rpm
	config is not available to yum and therefore a 3-way diff is not
	possible. Perhaps a future enhancement would be to use a .orig file,
	if present, as the merge base for meld.  When using vimdiff, the
	order has not changed. The first parameter to vimdiff is the "final:
	file i.e. the target file, and the second parameter is the "other"
	file i.e. the .rpmnew or .rpmsave file.

2011-06-10  James Antill <james@and.org>

	* repodiff.py: Huge speedup for large number of add/remove obsoletes
	processing.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* yumdownloader.py: Stops yumdownloader from attempting to download
	the same package twice if it is available in multiple repos. BZ
	711767

2011-06-10  James Antill <james@and.org>

	* docs/repodiff.1, repodiff.py: Add --compare-arch for the other
	side of the multilib. diff. BZ 710579.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* repodiff.py: Changed archlist set up to handle multilib. BZ 710579


2011-06-01  James Antill <james@and.org>

	* docs/repoquery.1: Add releasever and setopt docs. from yum to
	repoquery. BZ 638542.

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* yum-utils.spec: frelling pyyaml pkg named with weird case

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/puppetverify/puppetverify.conf,
	plugins/puppetverify/puppetverify.py, yum-utils.spec: add
	puppetverify plugin

2011-05-18  Frank Weltersbach <frankwe@live.com>

	* docs/package-cleanup.1: Fix typo in package-cleanup man page. BZ
	705617.

2011-05-17  Seth Vidal <skvidal@fedoraproject.org>

	* repoquery.py:     - fix rh bug: #705179     - fix fmt_queryformat not being happy
	    with new 'output' formats passed       everywhere else.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Allow files/provides to be used in
	yumdownloader.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Exit with code 2, if we can't download the
	packages asked for. BZ 699469.

2011-04-26  James Antill <james@and.org>

	* plugins/auto-update-debuginfo/auto-update-debuginfo.py: try/except
	the open, for the rpmdbv cached data (a-u-d plugin).

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add yum-debug-dump completion.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Mention filename arg in yum-debug-dump usage
	message.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Make yum-debug-dump.py executable for running
	in-place.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Use functions from recent bash-completion if
	available.

2011-04-06  James Antill <james@and.org>

	* repoquery.py, yum-groups-manager.py: Convert

show-changed-rco.py
#! /usr/bin/python -tt

#!/usr/bin/python -tt

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import yum

import os

from optparse import OptionParser
from optparse import SUPPRESS_HELP

version = "1.0.0"


def _get_npkgs(self, args):
    pkgs = []
    for arg in args:
        if (arg.endswith('.rpm') and (yum.misc.re_remote_url(arg) or
                                      os.path.exists(arg))):
            thispkg = yum.packages.YumUrlPackage(self, self.ts, arg)
            pkgs.append(thispkg)
        elif self.conf.showdupesfromrepos:
            pkgs.extend(self.pkgSack.returnPackages(patterns=[arg]))
        else:                
            try:
                pkgs.extend(self.pkgSack.returnNewestByName(patterns=[arg]))
            except yum.Errors.PackageSackError:
                pass
    return pkgs

def _get_opkgs(self, npkgs, old_packages):
    if old_packages:
        return _get_npkgs(self, old_packages)
    pkg_names = set((pkg.name for pkg in npkgs))
    return self.rpmdb.searchNames(pkg_names)

def _get_oreqs(pkg, reqs):
    oreqs = {}
    for req in reqs:
        (r,f,v) = req
        if r.startswith('rpmlib('):
            continue

        if r not in oreqs:
            oreqs[r] = set([req])
        else:
            oreqs[r].add(req)
    return oreqs

def _get_reqs(pkg, reqs, oreqs, self_prov_check=True):
    nreqs = set()
    creqs = set()

    for req in reqs:
        (r,f,v) = req
        if r.startswith('rpmlib('):
            continue

        if r in oreqs and req in oreqs[r]:
            continue

        if (r in pkg.provides_names or
            (r[0] == '/' and r in (pkg.filelist + pkg.dirlist +
                                   pkg.ghostlist))):
            if not f or pkg.checkPrco('provides', req):
                continue

        if r in oreqs:
            creqs.add(req)
        else:
            nreqs.add(req)

    return nreqs, creqs


def _print_reqs(yb, pkg, reqs, used_repos):
    out_reqs = {}
    for req in reqs:
        (r,f,v) = req
        seen = {}
        out_reqs[req] = []
        for pkg in sorted(yb.rpmdb.searchProvides(req)):
            key = (pkg.name, pkg.arch)
            if key in seen and not yb.conf.showdupesfromrepos:
                continue
            seen[key] = pkg
            out_reqs[req].append(pkg)
            used_repos.append(pkg.ui_from_repo)
        for pkg in sorted(yb.pkgSack.searchProvides(req)):
            key = (pkg.name, pkg.arch)
            if key in seen and not yb.conf.showdupesfromrepos:
                continue
            seen[key] = pkg
            out_reqs[req].append(pkg)
            used_repos.append(pkg.ui_from_repo)
    done = set()
    for req in sorted(out_reqs):
        if req in done:
            continue
        done.add(req)
        print " ", yum.misc.prco_tuple_to_string(req)
        for oreq in sorted(out_reqs):
            if oreq in done:
                continue
            if req == oreq:
                continue
            if out_reqs[oreq] == out_reqs[req]:
                print " ", yum.misc.prco_tuple_to_string(oreq)
                done.add(oreq)
        for pkg in out_reqs[req]:
            print "   ", pkg, pkg.ui_from_repo

def _print_sum(title, used_repos, ind=" ", end=' '):
    if not used_repos:
        return

    print title
    crepos = {}
    installed = 0
    available = 0
    for urepo in used_repos:
        if urepo not in crepos:
            crepos[urepo] = 0
        crepos[urepo] += 1
        if urepo[0] == '@':
            installed += 1
        else:
            available += 1

    if installed:
        print ind, "Installed:", installed
    for urepo in sorted(crepos):
        if urepo[0] != '@':
            continue
        print "%s%s" % (ind, "   "), urepo + ":", crepos[urepo]
    if available:
        print ind, "Available:", available
    for urepo in sorted(crepos):
        if urepo[0] == '@':
            continue
        print "%s%s" % (ind, "   "), urepo + ":", crepos[urepo]

    if end:
        print end


def main():
    parser = OptionParser(version = "Depcheck version %s" % version)
    parser.add_option("--releasever", default=None,
                      help="set value of $releasever in yum config and repo files")
    parser.add_option("--show-duplicates", action="store_true",
                      dest="show_dupes",
                      help="show all versions of packages")
    parser.add_option("--show-dupes", action="store_true",
                      help=SUPPRESS_HELP)
    parser.add_option("--repoid", action="append",
                      help="specify repoids to query, can be specified multiple times (default is all enabled)")
    parser.add_option("--enablerepo", action="append", dest="enablerepos",
                      help="specify additional repoids to query, can be specified multiple times")
    parser.add_option("--disablerepo", action="append", dest="disablerepos",
                      help="specify repoids to disable, can be specified multiple times")                      
    parser.add_option("--repofrompath", action="append",
                      help="specify repoid & paths of additional repositories - unique repoid and complete path required, can be specified multiple times. Example. --repofrompath=myrepo,/path/to/repo")
    parser.add_option("--old-packages", action="append",
                      help="packages to use to compare against, instead of installed")
    parser.add_option("--ignore-arch", action="store_true",
                      help="ignore arch when searching for old packages")
    parser.add_option("--skip-new", action="store_true",
                      help="skip packages without a matching old package")
    parser.add_option("-C", "--cache", action="store_true",
                      help="run from cache only")
    parser.add_option("-c", "--config", dest="conffile", default=None,
                      help="config file location")

    (opts, args) = parser.parse_args()


    yb = yum.YumBase()
    yb.preconf.releasever = opts.releasever
    if opts.conffile is not None:
        yb.preconf.fn = opts.conffile

    # setup the fake repos
    for repo in opts.repofrompath or []:
        tmp = tuple(repo.split(','))
        if len(tmp) != 2:
            yb.logger.error("Error: Bad repofrompath argument: %s" %repo)
            continue
        repoid,repopath = tmp
        if repopath[0] == '/':
            baseurl = 'file://' + repopath
        else:
            baseurl = repopath
        yb.add_enable_repo(repoid, baseurls=[baseurl],
                           basecachedir=yb.conf.cachedir)
        yb.logger.info("Added %s repo from %s" % (repoid, repopath))

    if opts.cache:
        yb.conf.cache = 1
    elif not yb.setCacheDir():
        yb.conf.cache = 1

    if opts.show_dupes:
        yb.conf.showdupesfromrepos = True

    if opts.repoid:
        found_repos = set()
        for repo in yb.repos.findRepos('*'):
            if repo.id not in opts.repoid:
                repo.disable()
            else:
                found_repos.add(repo.id)
                repo.enable()
        for not_found in set(opts.repoid).difference(found_repos):
            yb.logger.error('Repoid %s was not found.' % not_found)

    if opts.disablerepos:
        for repo_match in opts.disablerepos:
            for repo in yb.repos.findRepos(repo_match):
                repo.disable()

    if opts.enablerepos:    
        for repo_match in opts.enablerepos:
            for repo in yb.repos.findRepos(repo_match):
                repo.enable()

    npkgs = _get_npkgs(yb, args)
    opkgs = {}
    for pkg in sorted(_get_opkgs(yb, npkgs, opts.old_packages)):
        opkgs[(pkg.name, pkg.arch)] = pkg
        opkgs[pkg.name] = pkg

    for pkg in sorted(npkgs):
        opkg = None
        oreqs = {}
        oobss = {}
        ocons = {}
        if opts.ignore_arch:
            if pkg.name in opkgs:
                opkg = opkgs[pkg.name]
        elif (pkg.name, pkg.arch) in opkgs:
            opkg = opkgs[(pkg.name, pkg.arch)]

        if opkg is None and opts.skip_new:
            continue

        print "New-Package:", pkg, pkg.ui_from_repo

        if opkg is not None:
            print "Old-Package:", opkg, opkg.ui_from_repo

            oreqs = _get_oreqs(pkg, opkg.requires)
            ocons = _get_oreqs(pkg, opkg.conflicts)
            oobss = _get_oreqs(pkg, opkg.obsoletes)

        used_repos_reqs = []
        nreqs, creqs = _get_reqs(pkg, pkg.requires, oreqs)
        if nreqs:
            print "New-Requires:"
            _print_reqs(yb, pkg, nreqs, used_repos_reqs)
        if creqs:
            print "Modified-Requires:"
            _print_reqs(yb, pkg, creqs, used_repos_reqs)

        _print_sum("Dep-Requires-Repos:", used_repos_reqs)

        used_repos_cons = []
        nreqs, creqs = _get_reqs(pkg, pkg.conflicts, ocons)
        if nreqs:
            print "New-Conflicts:"
            _print_reqs(yb, pkg, nreqs, used_repos_cons)
        if creqs:
            print "Mod-Conflicts:"
            _print_reqs(yb, pkg, creqs, used_repos_cons)

        _print_sum("Dep-Conflicts-Repos:", used_repos_cons)

        used_repos_obss = []
        nreqs, creqs = _get_reqs(pkg, pkg.obsoletes, oobss)
        if nreqs:
            print "New-Obsoletes:"
            _print_reqs(yb, pkg, nreqs, used_repos_obss)
        if creqs:
            print "Mod-Obsoletes:"
            _print_reqs(yb, pkg, creqs, used_repos_obss)

        _print_sum("Dep-Obsoletes-Repos:", used_repos_obss)
        _print_sum("Dep-Repos:",
                   used_repos_reqs + used_repos_cons + used_repos_obss,
                   end='')


if __name__ == "__main__":
    yum.misc.setup_locale()
    main()
